spinalcordtoolbox/labels.py

```diff
@@ -464,3 +464,79 @@ def remove_other_labels_from_image(img: Image, labels: Sequence[int]) -> Image:
             logger.warning(f"Label {label} not found in input image!")
 
     return out
+
+
+def project_centerline(img: Image, ref: Image) -> Image:
+    """
+    Project an input image on the spinal cord centerline. This projection is obtained by iterating along
+    the centerline to identify the shortest distance with each referenced coordinates.
+    Typically, user inputs a segmentation image, and labels with disks position, and this function computes
+    the identification to the closest coordinates of each labels on the centerline.
+
+    :param img: segmentation
+    :param ref: reference labels
+    :returns: image with the new projected labels on the centerline
+    """
+    # Checking orientation
+    og_img_orientation = img.orientation
+    if img.orientation != "RPI":
+        img.change_orientation("RPI")
+
+    og_ref_orientation = ref.orientation
+    if ref.orientation != "RPI":
+        ref.change_orientation("RPI")
+
+    # Extract centerline from segmentation
+    _, arr_ctl, _, _ = get_centerline(img, param=ParamCenterline())
+    centerline = arr_ctl.T
+
+    # Extract referenced coordinates
+    coordinates_ref = ref.getNonZeroCoordinates(sorting='value')
+
+    # Create the output image
+    out = zeros_like(ref)
+
+    # Compute the shortest distance for each referenced points on the centerline
+    for point in coordinates_ref:
+        projection = project_point_on_line(point=np.array(list(point)), line=centerline)
+        value = projection[-1]
+        projection = np.rint(projection[:3]).astype(int)
+        if out.data[projection[0], projection[1], projection[2]] == 0:
+            out.data[projection[0], projection[1], projection[2]] = value
+        elif out.data[projection[0], projection[1], projection[2]] == value:
+            logger.warning("Two labels with the same value were projected on the same coordinate")
+        else:
+            # overwrite with the highest value because referenced points are sorted
+            logger.warning("Two labels were projected on the same coordinate, the highest value was kept")
+            out.data[projection[0], projection[1], projection[2]] = value
+
+    if out.orientation != og_img_orientation:
+        out.change_orientation(og_img_orientation)
+        img.change_orientation(og_img_orientation)
+
+    if ref.orientation != og_ref_orientation:
+        ref.change_orientation(og_ref_orientation)
+
+    return out
+
+
+def project_point_on_line(point, line):
+    """
+    Project the input point on the referenced line by finding the minimal distance
+
+    :param point: coordinates of a point and its value: point = numpy.array([x y z value])
+    :param line: list of points coordinates which compose the line
+    :returns: closest coordinate to the referenced point on the line: projected_point = numpy.array([X Y Z value])
```

Comment:
Is there a reason to weave `value` through this function? It seems like it would be simpler to just pass `np.array([x, y, z])` and return `np.array([X, Y, Z])`.

(The caller already has to manipulate the input to change it from a `Coordinate` instance to a numpy array, so it might as well do the extra processing before calling `project_point_on_line`. Plus, the caller then has to separate out the `value` from the result anyway.)